l10n_ch_base_bank/models/invoice.py

```diff
@@ -57,71 +59,71 @@ def _get_reference_type(self):
+                    bank_acc = invoice.partner_banks_to_show()
+                else:
+                    bank_acc = invoice.partner_bank_id
+                if not bank_acc:
+                    raise exceptions.ValidationError(
+                        _("Bank account shouldn't be empty, for ISR reference"
+                          " type, you can set it manually or set appropriate"
+                          " payment mode.")
+                    )
+                if (not bank_acc._is_qr_iban()
+                        and (invoice.currency_id.name == 'CHF'
+                             and not bank_acc.l10n_ch_isr_subscription_chf)
+                        or (invoice.currency_id.name == 'EUR'
+                            and not bank_acc.l10n_ch_isr_subscription_eur)):
+                    raise exceptions.ValidationError(
+                        _("Bank account must contain a subscription number for"
+                          " ISR reference type.")
+                    )
         return True
 
-    @api.multi
     def _is_isr_reference(self):
+        """Check if the communication is a valid ISR reference
+        e.g.
+        210000000003139471430009017
+        21 00000 00003 13947 14300 09017
+        This is used to determine SEPA local instrument
         """
-        Function to validate a ISR reference like :
-        0100054150009>132000000000000000000000014+ 1300132412>
-        The validation is based on l10n_ch
-        """
-        if not self.reference:
-            raise exceptions.ValidationError(
-                _('ISR Reference is required')
-            )
-        # In this case
-        # <010001000060190> 052550152684006+ 43435>
-        # the reference 052550152684006 do not match modulo 10
-        #
-        if (mod10r(self.reference[:-1]) != self.reference and
-                len(self.reference) == 15):
-            return True
-        #
-        if mod10r(self.reference[:-1]) != self.reference:
-            raise exceptions.ValidationError(
-                _('Invalid ISR Number (wrong checksum).')
-            )
-
-    @api.constrains('reference')
-    def _check_isr(self):
-        """ Do the check only for invoice with reference_type = 'isr' """
-        for invoice in self:
-            if invoice.reference_type == 'isr':
-                invoice._is_isr_reference()
-        return True
+        if self.reference_type == 'isr':
+            if (mod10r(self.reference[:-1]) != self.reference and
+                    len(self.reference) == 15):
+                return True
+            if mod10r(self.reference[:-1]) != self.reference:
+                raise exceptions.ValidationError(
+                    _('Invalid ISR Number (wrong checksum).')
+                )
+        elif self.reference_type == 'qrr':
+            if not self.reference:
+                return False
+            if re.match(r'^(\d{27}|\d{2}( \d{5}){5})$', self.reference):
```

Comment:
I think there might be an issue with this regular expression. Can you explain me what is the purpose of the expression ?

Reply:
Hi @dcoopennet, the regular expression is got from the v12 of the module, but the reg ex try to match a valid QRR Iban code.
The regular expression is the same as here: https://github.com/OCA/l10n-switzerland/blob/12.0/l10n_ch_base_bank/models/invoice.py#L93